### ISSUE TITLE
Query Endpoint information/ user data in server accept callback or later

### ISSUE THREAD

**Issue by Akshay-Venkatesh** [openucx/ucx#3152]
Hi,

Upon successful accept at listener, today I have an opaque handle for the ep connected to the peer. Is it possible to query information about the newly connected EP? Information such as remote hostname, pid, EP pointer is what I'm looking to extract. I don't know if these are possible today. It would useful to be able to pass user data and user data length to ep_create in server-client connection mode, and query this upon accept. 

For background, this requirement comes from wanting to generate EP-specific tag for ucp_tag_recv_nb/probe_nb. Today, it's possible to exchange this information separately once the connections are established but having user data as part of ep_create could simplify this. Lmk your thoughts. Thanks

**Comment by evgeny-leksikov** [comment]
Do you generate TAG to identify EP on recv side? Why just do not use stream API?`ucp_stream_worker_poll` returns EPs with arrived data, and then do `ucp_stream_recv_nb` on specific EP. If you need something "message oriented" you can put size to header and recv msg by 2 calls in whorse case: `ucp_stream_recv_data_nb` to get header and some portion of data and do `ucp_stream_recv_nb(... , UCP_STREAM_RECV_FLAG_WAITALL)` to receive the rest of the data in single completion,

According to your request - at the moment when you have UCP EP on server side it may be not really connected to client. Until wireup protocol is completed we have only small amount of private data from the client, it not always even fit our needs. It comes from rdma cm limitations and this is why we start "generic" wireup. What if we add "connected" callback to `ucp_ep_create` for client/server EPs (may it be useful for other EP types?) which would provide details about the peer? This way, we can provide the EP info to user and do not consume internal EP memory @yosefe WDYT?

**Comment by Akshay-Venkatesh** [comment]
> Do you generate TAG to identify EP on recv side? Why just do not use stream API?`ucp_stream_worker_poll` returns EPs with arrived data, and then do `ucp_stream_recv_nb` on specific EP. If you need something "message oriented" you can put size to header and recv msg by 2 calls in whorse case: `ucp_stream_recv_data_nb` to get header and some portion of data and do `ucp_stream_recv_nb(... , UCP_STREAM_RECV_FLAG_WAITALL)` to receive the rest of the data in single completion,
> 
Actually, instead of `stream_recv_nb` can I not just use `tag_recv_nb` after accept? (BTW, this is currently what I'm doing). I use `tag_recv_nb` to receive some external information about the connection. I use that external information as a tag for the rest of transfers on that newly accepted connection. I prefer to avoid this approach because I have to wait for the completion of the `tag_recv_nb` by polling after accept before any new exchanges. Instead, if I have a field in `ep` such as `ep.user_data` (which is received as part of connection establishment), I can query this and avoid `tag_recv_nb` and `worker_progress` call. It purely simplifies work and I'm likely going to piggyback on some exchanges that are likely taking place to exchange transport information as part of client-server connection establishment.

> According to your request - at the moment when you have UCP EP on server side it may be not really connected to client. Until wireup protocol is completed we have only small amount of private data from the client, it not always even fit our needs. It comes from rdma cm limitations and this is why we start "generic" wireup. What if we add "connected" callback to `ucp_ep_create` for client/server EPs (may it be useful for other EP types?) which would provide details about the peer? This way, we can provide the EP info to user and do not consume internal EP memory @yosefe WDYT?

This would be useful as well, IMO. This allows both the client side and server side to exchange connection information or some external user information about each other.

**Comment by evgeny-leksikov** [comment]
The difference between `tag` and `stream` recvs is a context object. Tag matching context is a worker, so you will receive messages from all clients here and need to identify the client by TAG and guarantee no collisions generating it on the server or using some hash function and extra user info as an argument. I guess the last thing is what you are trying to do.
Stream recv context is an endpoint. Posting `stream_recv_nb` after `accept` guarantees you that the data will be received from the specific client and you can handle it immediately or later in completion callback.

> This allows both the client side and server side to exchange connection information or some external user information about each other.

Yes, I would like just to understand the case and why you really need tag matching here and why stream API doesn't match your needs.

**Comment by Akshay-Venkatesh** [comment]
@evgeny-leksikov my apologies but you're right and I missed noticing that stream_recv_nb has an EP parameter. Stream API definitely suffices.

However, we plan to send and receive from cuda buffers. Isn't there a performance limitation with stream operations in cuda buffers today?

**Comment by evgeny-leksikov** [comment]
Today we have implementation of stream API over UCT AMs, there is no rdnv yet, but cuda memory should be supported for AM bcopy/zcopy at least. @bureddy do you have any perf numbers of stream send/recv vs tag API on cuda memory?

**Comment by bureddy** [comment]
the cuda perf is very bad because of AM bcopy/zcopy. And it could not use cuda perf protocols ( GDR, cudaIPC) because of no rndv.

**Comment by shamisp** [comment]
Does it mean that it has to bounce GPU->CPU->HCA , instead of GPU->HCA ?

**Comment by Akshay-Venkatesh** [comment]
@shamisp I believe yes. As @bureddy mentioned, I see that `ucp_stream_send_am_short` or `ucp_request_send_start` is called from stream send ops which map to UCP_REQUEST_SEND_PROTO_BCOPY_AM or UCP_REQUEST_SEND_PROTO_ZCOPY_AM protocols. 

@evgeny-leksikov A few months ago, I ran into the CUDA stream performance limitation when we had started with the use of stream API for ucx python bindings. At the time, using stream API with client-server connection model made sense. Noting the limitation, we moved to using tag_recv/tag_send with client-server instead. For now, we have something that works with tag_recv/tag_send but I brought this up because: 1. I ignored that stream_recv accepts ep as a parameter 2. I thought obtaining user data at accept could be useful. I get around this today by explicitly calling send at client and recv at server immediately after connection. 
With this discussion, I'm not so sure if the second step is too painful as a placeholder until stream send/recv becomes performant for CUDA buffers. I'm ok with closing the issue at this point.
